### PR TITLE
fix: add support for nested index routes (fixes #281)

### LINKED ISSUE #281: preflight 404 on prefixed index routes in nested/scopted cors setups
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.x

### Plugin version

8.4.2

### Node.js version

16.X

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

13.5

### Description

Preflight request routing is currently broken for index routes in prefixed plugins with scoped cors routes.

I have a setup right now where I instantiate CORS within scoped route plugins - think differing cors rules for `/public` and `/private`.

Within each of these plugins, I setup my cors plugin, and have an index route - so within the `/public` plugin I add a `/` route.

However, this index route has no preflight support because the options route added by fastify-cors is scoped to `/*` (previously, this was just `*`, but that broke some other setups)

A possible workaround would be to register another options handler on `/`.

### Steps to Reproduce

Repro: https://github.com/GlenTiki/fastify-cors-bug

```js
const Fastify = require("fastify");
const cors = require("@fastify/cors");

const server = Fastify();

const prefixPlugin = async (app) => {
  app.register(cors);
 
  app.get("/", async (request, reply) => {
    return { hello: "world" };
  });
}
server.register(prefixPlugin, {prefix: "/example"});

const assert = require("assert");

async function runTest() {
  await server.ready();
  console.log(server.printRoutes())

  const res = await server.inject({
    url: "/example",
    method: "OPTIONS",
    headers: {
      Accept: "*/*",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
      "Access-Control-Request-Headers": "content-type",
      "Access-Control-Request-Method": "POST",
      "Cache-Control": "no-cache",
      Connection: "keep-alive",
      Host: "localhost:3001",
      Origin: "http://localhost:8080",
      Pragma: "no-cache",
      Referer: "http://localhost:8080/",
      "Sec-Fetch-Dest": "empty",
      "Sec-Fetch-Mode": "cors",
      "Sec-Fetch-Site": "same-site",
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36 Edg/119.0.0.0",
    },
  });

  console.log(res.statusCode);
  console.log(res.headers);

  // works with "@fastify/cors": "8.4.1"
  // broken with "@fastify/cors": "8.4.2"
  assert(res.statusCode === 204)
}

runTest();
```

### Expected Behavior

Preflight requests should not 404 on the index route.

### PR DESCRIPTION
**Pull request by GlenTiki**
Fixes #281. I'll comment on code to clarify reasoning.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] documentation is changed or added

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -70,17 +70,21 @@ function fastifyCors (fastify, opts, next) {
...
     }
 
     reply.send()
-  })
+  }
+
+  // This route enables fastify to accept preflight requests on all nested routes, as well as the index route in prefixed setups
+  // https://github.com/fastify/fastify-cors/issues/281
+  fastify.options('/*', handlerOptions, handler)
+  if (fastify.prefix) { fastify.options('/', handlerOptions, handler) }
```
Comment (GlenTiki) [inline index.js:87]: without doing this if check existing tests break - this literally just fixes the case for the test I added.

I think on root index routes, fastify applies `/*` to the index route, but on prefixed routes, it doesn't?

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Uzlopak — CHANGES_REQUESTED** [review]
I am not at my PC but reading this i think the solution is wrong in case fastify.prefix is set. I mean, yes the test passes but because only this one case was patched. 

I assume that you need to do 
```js
fastify.options(fastify.prefix + '*', handlerOptions, handler)
```
To fix it properly.

Blocking the merging.

**Comment by GlenTiki** [comment]
@Uzlopak when you’re back at a desk can you take a look again and give further guidance on how to fix this? I tried some variations on your suggestion, but that fails the added test case.

**Review by Uzlopak — COMMENTED** [review]
I reviewed this again.

Yes, your assumption regarding index seems to be right. But it feels like this is an error in upstream, either find-my-way or fastify itself. 

So unfortunately I can not give you more hints. 

Something is odd. Well this is my gut feeling. 

@mcollina 
@ivan-tymoshenko

@GlenTiki wrote;
> I think on root index routes, fastify applies /* to the index route, but on prefixed routes, it doesn't?


Can this be a bug in upstream?

**Comment by ivan-tymoshenko** [comment]
It looks like a bug in fastify for me.

**Comment by GlenTiki** [comment]
I tested fastify with some simple tests which are passing (attached below). Meaning the issue lies in this repo.

I'm now digging in to see if the error lies in the hook this repo adds, but, given this solves the test case, it's an alternate valid solution 😅 

<details>
<summary>Expand to see test</summary>

```js
'use strict'

const t = require('tap')
const test = t.test
const Fastify = require('fastify')

test('Routing to the wildcard route works the same for prefixed or root route', async t => {
    t.plan(3)

    const fastify = Fastify()

    const handler = async (req, reply) => {
        return { custom: "body" }
    }

    fastify.route({ method: 'OPTIONS', url: '/*', handler })
    fastify.register(async (scope, opts) => {
        scope.route({ method: 'OPTIONS', url: '/*', handler })
    }, { prefix: '/prefix' })

    await fastify.ready()

    const result = (await fastify.inject({ method: "OPTIONS", path: '/' })).payload

    t.equal(result, JSON.stringify({ custom: "body" }))
    t.equal(result, (await fastify.inject({ method: "OPTIONS", path: '/prefix' })).payload)
    t.equal(result, (await fastify.inject({ method: "OPTIONS", path: '/prefix/' })).payload)
})


test('Routing to the wildcard route works the same for nested prefixed or root route', async t => {
    t.plan(3)

    const fastify = Fastify()

    const handler = async (req, reply) => {
        return { custom: "body" }
    }

    const nestedWildcardPlugin = async (scope, opts) => {
        scope.route({ method: 'OPTIONS', url: '/*', handler })
    }

    fastify.register(nestedWildcardPlugin)
    fastify.register(async (scope, opts) => {
        scope.register(nestedWildcardPlugin)
    }, { prefix: '/prefix' })

    await fastify.ready()

    const result = (await fastify.inject({ method: "OPTIONS", path: '/' })).payload

    t.equal(result, JSON.stringify({ custom: "body" }))
    t.equal(result, (await fastify.inject({ method: "OPTIONS", path: '/prefix' })).payload)
    t.equal(result, (await fastify.inject({ method: "OPTIONS", path: '/prefix/' })).payload)
})

```

</details>

**Review by climba03003 — CHANGES_REQUESTED** [review]
You should be using the similar approach for 404 catcher in `fastify`.
https://github.com/fastify/fastify/blob/af8bd46a7b1d18c68558f07e020180f5e86922b5/lib/fourOhFour.js#L160-L161C11

**Comment by Uzlopak** [comment]
@climba03003 

I tested that and it didnt work. Can you provide a suggestion to fix this issue?

**Comment by climba03003** [comment]
I would like to revert https://github.com/fastify/fastify-cors/pull/278 after some investigation.
We cannot replace `*` with `/*` directly without sacrifice the ability to to replace response for `options` route.

We need another approach to fix the issue mentioned inside #278

**Comment by Fdawgs** [comment]
> I would like to revert https://github.com/fastify/fastify-cors/pull/278 after some investigation.
> 
> We cannot replace `*` with `/*` directly without sacrifice the ability to to replace response for `options` route.
> 
> 
> 
> We need another approach to fix the issue mentioned inside #278

Sounds good!

**Comment by mcollina** [comment]
@climba03003 can you open a PR and add a test to prevent the regression in the future?

**Pull request closed without merging**